tutorials/pde/2_advection_diffusion.py

```diff
@@ -0,0 +1,178 @@
+    - \\alpha \int_\\Omega (\\nabla \\psi ) \\cdot (\\nabla p ) \\, dv \n
+    &- \int_{\\partial \\Omega} \\mathbf{n} \\cdot ( \\psi \\, p \\, \\mathbf{u}) \\, da
+    + \int_\\Omega \\big ( \\nabla \\cdot (\mathbf{u} \\, \\psi ) \\big) \\, p \\, dv
+    + \\psi \\, q
+
+Since the flux on the faces is zero at the boundaries, we can eliminate
+the first term on the right-hand side. By defining :math:`\\mathbf{n \\cdot u} = 0`
+on the boundary, the third term on the right-hand side is zero. By evaluating
+the inner products according to the finite volume approach we obtain:
+
+.. math::
+    \\mathbf{\\psi^T M_c p_t} = - \\alpha \\mathbf{\\psi^T G^T M_f G \\, p}
+    + \\mathbf{\\psi^T A_{cf}^T} \\, \\textrm{diag} ( \\mathbf{u} ) \\mathbf{D^T \\, M_c \\, p}
+    + \\mathbf{\\psi^T M_c s}
+
+where :math:`\\mathbf{\\psi}`, :math:`\\mathbf{p}` and :math:`\\mathbf{p_t}`
+live at cell centers and :math:`\\mathbf{u}` lives on faces. :math:`\\mathbf{D}`
+is the discrete divergence operator. :math:`\\mathbf{G}`
+is the discrete gradient operator. :math:`\\mathbf{M_c}` and
+:math:`\\mathbf{M_f}` are the cell center and cell faces inner product matricies,
+respectively. :math:`\\mathbf{A_{cf}}` averages from cell centers to cell
+faces.
+
+By eliminating :math:`\\psi^T` and multiplying both sides by
+:math:`\\mathbf{M_c^{-1}}` we obtain:
+
+.. math::
+    \\mathbf{p_t} = - \\mathbf{M} \\mathbf{p} + \\mathbf{s}
+
+where
+
+.. math::
+    \\mathbf{M} = \\alpha \\mathbf{M_c^{-1} G^T M_f G}
+    - \\mathbf{M_c^{-1} A_{cf}^T} \\, \\textrm{diag}(\\mathbf{u}) \\mathbf{D^T M_c}
+
+For the example, we will discretize in time using backward Euler. This results
+in the following system which must be solve at every time step :math:`k`.
+Where :math:`\\Delta t` is the step size:
+
+.. math::
+    \\big [ \\mathbf{I} + \\Delta t \\, \\mathbf{M} \\big ] \\mathbf{p}^{k+1} =
+    \\mathbf{p}^k + \\Delta t \\, \\mathbf{s}
+    
+
+"""
+
+###################################################
+#
+# Import Packages
+# ---------------
+#
+# Here we import the packages required for this tutorial.
+#
+
+from discretize import TensorMesh
+from pymatsolver import SolverLU
+import matplotlib.pyplot as plt
+import numpy as np
+from discretize.utils import sdiag
+
+###############################################
+#
+# Solving the Problem
+# -------------------
+#
+
+# Create a tensor mesh
+h = np.ones(75)
+mesh = TensorMesh([h, h], 'CC')
+
+# Define diffusivity constant
+a = 25
+
+# Define velocity vector u = [ux, uy] = [10, 0] on faces
+ux = 10*np.ones((mesh.nFx))
+uy = np.zeros((mesh.nFy))
+u = np.r_[ux, uy]
+
+# Define source term diag(v)*s
+xycc = mesh.gridCC
+k1 = (xycc[:, 0]==-25) & (xycc[:, 1]==25)   # source at (-25, 25)
+k2 = (xycc[:, 0]==25) & (xycc[:, 1]==-25)   # sink at (25, -25)
+
+s = np.zeros(mesh.nC)
+s[k1] = 10
+s[k2] = 10
+
+# Define system M
+Acf = mesh.aveCC2F
+Mc_inv = sdiag(1/mesh.vol)
```

Comment:
We do not need `Mc_inv`

Reply:
It just seemed easier than using Mc and doing a solve. I know it is easy and fast because it is diagonal. I can change it if it's really confusing.